google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/sdk/ManagedCloudSdkService.java

```diff
@@ -16,35 +16,218 @@
+ *
+ * <p>Some methods ({@link #install()} do their work on background thread but listeners methods are
+ * always called on EDT thread. This class is not thread-safe, must be used on UI thread.
  */
-// TODO(ivanporty) implementation coming in the next PR
 public class ManagedCloudSdkService implements CloudSdkService {
+  private Logger logger = Logger.getInstance(ManagedCloudSdkService.class);
+
+  private ManagedCloudSdk managedCloudSdk;
+
+  private SdkStatus sdkStatus = SdkStatus.NOT_AVAILABLE;
+
+  private final Collection<SdkStatusUpdateListener> statusUpdateListeners = Lists.newArrayList();
+
+  private volatile ListenableFuture<Path> runningInstallationJob;
+
+  @Override
+  public void activate() {
+    // TODO track event that custom SDK is activated and used.
+
+    install();
+  }
 
   @Nullable
   @Override
   public Path getSdkHomePath() {
-    return null;
+    switch (sdkStatus) {
+      case READY:
+        return managedCloudSdk.getSdkHome();
+      default:
+        return null;
+    }
   }
 
   @Override
-  public void setSdkHomePath(String path) {}
+  public void setSdkHomePath(String path) {
+    /* unsupported, to be removed. */
+  }
 
   @Override
   public SdkStatus getStatus() {
-    return SdkStatus.NOT_AVAILABLE;
+    return sdkStatus;
   }
 
   @Override
-  public boolean installAutomatically() {
+  public boolean install() {
+    // check if installation is already running first, to prevent multiple jobs running.
+    if (runningInstallationJob == null) {
+      runningInstallationJob =
+          ThreadUtil.getInstance().executeInBackground(this::installSynchronously);
+      Futures.addCallback(
+          runningInstallationJob,
+          new FutureCallback<Path>() {
+            @Override
+            public void onSuccess(Path path) {
+              logger.info("Managed Google Cloud SDK successfully installed at: " + path);
+
+              runningInstallationJob = null;
+            }
+
+            @Override
+            public void onFailure(Throwable t) {
+              if (t instanceof InterruptedException) {
+                logger.info("Managed Google Cloud SDK installation cancelled.");
+              }
+
+              runningInstallationJob = null;
+            }
+          });
+    }
+
     return true;
   }
 
   @Override
-  public void addStatusUpdateListener(SdkStatusUpdateListener listener) {}
+  public void addStatusUpdateListener(SdkStatusUpdateListener listener) {
+    statusUpdateListeners.add(listener);
+  }
+
+  @Override
+  public void removeStatusUpdateListener(SdkStatusUpdateListener listener) {
+    statusUpdateListeners.remove(listener);
+  }
+
+  public void cancelInstall() {
+    if (runningInstallationJob != null) {
+      runningInstallationJob.cancel(true);
+    }
+  }
+
+  @VisibleForTesting
+  ManagedCloudSdk createManagedSdk() throws UnsupportedOsException {
+    return ManagedCloudSdk.newManagedSdk();
+  }
+
+  @VisibleForTesting
+  void setLogger(Logger logger) {
```

Comment:
Take a look at `UsefulTestCase`. Could you, instead of exposing a setter like this, do `Logger.setFactory` and pass in your own implementation if you need to?

Reply:
Also I'm wondering about the general approach of asserting log statements here. For the "operation successful" case for example, wouldn't it be better to assert that success occurred instead (the correct callback was called or whatever)?

Reply:
You might be right that it's too close testing even including logs. Success and failure of installation are tested via public APIs a lot in other unit tests (i.e. `activate_service_sdkInstalled_status_ready()` etc.). We can remove logging checks completely and check for messages when UI arrives.

Reply:
As for `Logger.setFactory`, should we consider moving this as a default to `CloudToolsRule`?

Reply:
do you mean so that it sets the `TestLoggerFactory`? I think that would make sense

Reply:
Filed in a separate issue/PR.